server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java

```diff
@@ -0,0 +1,185 @@
+        scriptScoreFunctionBuilder = in.readNamedWriteable(ScriptScoreFunctionBuilder.class);
+        minScore = in.readOptionalFloat();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(query);
+        out.writeNamedWriteable(scriptScoreFunctionBuilder);
+        out.writeOptionalFloat(minScore);
+    }
+
+    /**
+     * Returns the query builder that defines which documents the script_score query will be executed on.
+     */
+    public QueryBuilder query() {
+        return this.query;
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(QUERY_FIELD.getPreferredName());
+        query.toXContent(builder, params);
+        builder.field(SCRIPT_FIELD.getPreferredName(), scriptScoreFunctionBuilder.getScript());
+        if (minScore != null) {
+            builder.field(MIN_SCORE_FIELD.getPreferredName(), minScore);
+        }
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    public ScriptScoreQueryBuilder setMinScore(float minScore) {
+        this.minScore = minScore;
+        return this;
+    }
+
+    public Float getMinScore() {
+        return this.minScore;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected boolean doEquals(ScriptScoreQueryBuilder other) {
+        return Objects.equals(this.query, other.query) &&
+            Objects.equals(this.scriptScoreFunctionBuilder, other.scriptScoreFunctionBuilder) &&
+            Objects.equals(this.minScore, other.minScore) ;
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(this.query, this.scriptScoreFunctionBuilder, this.minScore);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        ScriptScoreFunction function = (ScriptScoreFunction) scriptScoreFunctionBuilder.toFunction(context);
+        Query query = this.query.toQuery(context);
+        return new ScriptScoreQuery(query, function, minScore);
+    }
+
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        QueryBuilder newQuery = this.query.rewrite(queryRewriteContext);
+        if (newQuery != query) {
+            ScriptScoreQueryBuilder newQueryBuilder = new ScriptScoreQueryBuilder(newQuery, scriptScoreFunctionBuilder);
+            newQueryBuilder.setMinScore(minScore);
+            return newQueryBuilder;
+        }
+        return this;
+    }
+
+    @Override
+    protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
+        InnerHitContextBuilder.extractInnerHits(query(), innerHits);
+    }
+
+    private static ConstructingObjectParser<ScriptScoreQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(NAME, false,
```

Comment:
nit: Could we put the static blocks, methods and constants at the top of the class?